// @zyb-data/stats-mp/dist/lib/h5/page.js
import { __assign } from "tslib";
import { DataType, EventDataType, BuiltInName } from '@zyb-data/stats-core';
import { D } from '../core/debug';
import { browserStatus } from './browser';
var ErrCPDoNotExist = '$cp do not exist.';
// const axios = require('axios')
var TargetVisionStatus;
(function (TargetVisionStatus) {
    TargetVisionStatus[TargetVisionStatus["Show"] = 0] = "Show";
    TargetVisionStatus[TargetVisionStatus["Hide"] = 1] = "Hide";
})(TargetVisionStatus || (TargetVisionStatus = {}));
var loc = typeof window === 'object' ? window.location : { origin: '', pathname: '' };
var StatsPage = /** @class */ (function () {
    function StatsPage($sm) {
        this.$sm = $sm;
        this.events = ['click'];
        this.route = loc.origin + loc.pathname;
        this.query = {};
        this.visionCheckTimerID = null;
        this.visionCheckTimeout = 250;
        this.visionCache = {};
    }
    /**
     * 初始化
     */
    StatsPage.prototype.init = function () {
        var _this = this;
        // 解析url
        this.parseUrl();
        // 设置Context 数据
        this.setContext();
        // 初始化视图检查
        this.visionCheck();
        this.events.forEach(function (event) {
            // 在捕获阶段触发，防止stopPropagation
            document.body.addEventListener(event, _this.eventHandler(), true);
        });
        // 页面监听
        this.pageCheck();
    };
    StatsPage.prototype.pageCheck = function () {
        var _this = this;
        var hidden = document.hidden;
        var res = browserStatus();
        if (hidden) {
            this.onHide();
        }
        else {
            this.onShow();
        }
        // 添加监听器
        document.addEventListener(res.visibilityChange, function () {
            status = document[res.state];
            if (document[res.state] == 'hidden') {
                _this.onHide();
            }
            else {
                _this.onShow();
            }
        }, false);
    };
    /**
     * 页面呈现
     */
    StatsPage.prototype.onShow = function () {
        try {
            this.$sm.$composer.composeBuiltin({
                name: BuiltInName.PageShow,
                pageUUID: this.route,
            });
            this.startTime = +new Date;
        }
        catch (err) {
            D.error(err);
        }
        D.log('页面展现（onShow）');
    };
    /**
     * 页面隐藏
     */
    StatsPage.prototype.onHide = function () {
        try {
            // @ts-ignore
            var eventDuration = parseInt((+new Date - this.startTime) / 1000);
            this.$sm.$composer.composeBuiltin({
                name: BuiltInName.PageHide,
                pageUUID: this.route,
                event_duration: eventDuration || 1
            });
        }
        catch (err) {
            D.error(err);
        }
        D.log('页面展现（onHide）');
    };
    StatsPage.prototype.setRouteRule = function (rule) {
        this.rule = rule;
    };
    /**
     * 视图变化
     */
    StatsPage.prototype.visionCheck = function () {
        var $this = this;
        var $cp = this.$cp;
        var pid = this.rule ? this.rule : this.route;
        var $sm = this.$sm;
        var check = function () {
            var _this = this;
            var _a;
            if (pid) {
                var suspects = Array.from(new Set($sm.$composer
                    .getSuspecteds(pid, EventDataType.viewShow)
                    .concat($sm.$composer.getSuspecteds(pid, EventDataType.viewHide))));
                (_a = $sm.$platform) === null || _a === void 0 ? void 0 : _a.getVisibleTargets(suspects).then(function (_a) {
                    var hit = _a.hit, unhit = _a.unhit;
                    hit.forEach(function (target) {
                        var targetInCache = $this.visionCache[target.id + "-" + target.zsQueryIndex];
                        if (targetInCache === undefined || targetInCache.status !== TargetVisionStatus.Show) {
                            var eventData = {
                                ty: EventDataType.viewShow,
                                id: target.id,
                                pid: _this.rule ? _this.rule : pid,
                                t: +new Date(),
                                ext: __assign({}, target.dataset)
                            };
                            if ($cp) {
                                $sm.$dm.setData(DataType.Event, eventData, $cp);
                                $sm.$composer.compose(eventData, $cp);
                            }
                            else {
                                throw new Error(ErrCPDoNotExist);
                            }
                            $this.visionCache[target.id + "-" + target.zsQueryIndex] = {
                                target: target,
                                status: TargetVisionStatus.Show
                            };
                        }
                    });
                    unhit.forEach(function (target) {
                        var targetInCache = $this.visionCache[target.id + "-" + target.zsQueryIndex];
                        if ((targetInCache === null || targetInCache === void 0 ? void 0 : targetInCache.status) === TargetVisionStatus.Show) {
                            // D.log('真实展现（hide）', target)
                            var eventData = {
                                ty: EventDataType.viewHide,
                                id: target.id,
                                pid: _this.rule ? _this.rule : pid,
                                t: +new Date(),
                                ext: __assign({}, target.dataset)
                            };
                            if ($cp) {
                                $sm.$dm.setData(DataType.Event, eventData, $cp);
                                $sm.$composer.compose(eventData, $cp);
                            }
                            else {
                                throw new Error(ErrCPDoNotExist);
                            }
                            $this.visionCache[target.id + "-" + target.zsQueryIndex] = {
                                target: target,
                                status: TargetVisionStatus.Hide
                            };
                        }
                    });
                });
            }
            else {
                D.log('异步的隐藏判断，上下文丢失了，用缓存上下文操作');
            }
            $this.visionCheckTimerID = setTimeout(check, $this.visionCheckTimeout);
        };
        var visionCheck = function () {
            $this.clearVisionCheckTimer();
            check();
        };
        visionCheck();
    };
    StatsPage.prototype.clearVisionCheckTimer = function () {
        if (this.visionCheckTimerID !== null) {
            clearTimeout(this.visionCheckTimerID);
        }
    };
    StatsPage.prototype.eventHandler = function () {
        var $this = this;
        var $sm = this.$sm;
        var eventHandler = function (event) {
            var _this = this;
            var _a;
            try {
                var pid_1 = $this.rule ? $this.rule : $this.route;
                var target = event.target;
                var suspects = $sm.$composer.getSuspecteds(pid_1, EventDataType.click);
                (_a = $sm.$platform) === null || _a === void 0 ? void 0 : _a.getHitTargets(0, 0, suspects, target).then(function (tar) {
                    if (!tar) {
                        return;
                    }
                    D.log('点击检查', tar);
                    var ext = __assign({}, tar.dataset);
                    try {
                        if (ext.zsIdx !== undefined && ext.zsList !== undefined) {
                            ext.zsItem = ext.zsList[ext.zsIdx] || {};
                        }
                    }
                    catch (e) {
                        D.log(e);
                    }
                    var eventData = {
                        ty: EventDataType.click,
                        id: tar.id,
                        pid: _this.rule ? _this.rule : pid_1,
                        t: +new Date(),
                        ext: ext
                    };
                    if ($this.$cp) {
                        $sm.$dm.setData(DataType.Event, eventData, $this.$cp);
                        $sm.$composer.compose(eventData, $this.$cp);
                    }
                    else {
                        throw new Error(ErrCPDoNotExist);
                    }
                });
            }
            catch (err) {
                D.error(err);
            }
        };
        return eventHandler;
    };
    StatsPage.prototype.parseUrl = function (route) {
        this.route = route ? route : location.origin + location.pathname;
        if (this.$cp) {
            this.$cp.pid = this.route;
        }
    };
    StatsPage.prototype.setContext = function () {
        this.$cp = this.$sm.$dm.appendDataPage(this.route);
        this.$sm.$dm.setData(DataType.Context, this.query, this.$cp);
    };
    return StatsPage;
}());
export { StatsPage };
