// hybrid-sdk-common/rollup.config.js
import typescript from 'rollup-plugin-typescript2'
import { terser } from "rollup-plugin-terser"
// import camelCase from 'lodash.camelcase'
import resolve from 'rollup-plugin-node-resolve'
import commonjs from 'rollup-plugin-commonjs';
import json from 'rollup-plugin-json';

import pkg from './package.json'

export default {
    input: 'src/index.ts',
    output: [
        {
            file: pkg.main || './lib/index.cjs.js',
            format: 'cjs',
        },
        {
            file: pkg.module || "./lib/index.es.js",
            format: 'es',
        },
        {
            file: pkg.browser || "./lib/index.iife.js",
            format: 'iife',
            name: 'HybridSdkCommon',
        }
    ],
    external: [
        ...Object.keys(pkg.peerDependencies || {}),
        'axios'
    ],
    plugins: [
        commonjs(),
        typescript({
            typescript: require('typescript'),
        }),
        terser(),
        resolve({ browser:true }),
        json()
    ],
}
