// @zyb-data/stats-core/dist/stats-core.umd.js
(function (global, factory) {
  typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports) :
  typeof define === 'function' && define.amd ? define(['exports'], factory) :
  (factory((global.statsCore = {})));
}(this, (function (exports) { 'use strict';

  /**
   * 事件类型枚举
   */
  (function (EventDataType) {
      /**
       * 进入页面
       */
      EventDataType["pageIn"] = "pageIn";
      /**
       * 离开页面
       */
      EventDataType["pageOut"] = "pageOut";
      /**
       * 控件显示
       */
      EventDataType["viewShow"] = "viewShow";
      /**
       * 控件隐藏
       */
      EventDataType["viewHide"] = "viewHide";
      /**
       * 点击
       */
      EventDataType["click"] = "click";
      /**
       * 请求开始
       */
      EventDataType["reqStart"] = "reqStart";
      /**
       * 请求成功
       */
      EventDataType["reqSucc"] = "reqSucc";
      /**
       * 请求失败
       */
      EventDataType["reqFail"] = "reqFail";
      /**
       * 分享
       */
      EventDataType["share"] = "share";
  })(exports.EventDataType || (exports.EventDataType = {}));
  (function (Comparator) {
      /**
       * 等于
       */
      Comparator["Equal"] = "eq";
      /**
       * 不等于
       */
      Comparator["NEQ"] = "neq";
      /**
       * 大于
       * @remark 仅支持数字
       */
      Comparator["GT"] = "gt";
      /**
       * 小于
       * @remark 仅支持数字
       */
      Comparator["LT"] = "lt";
      /**
       * 字符串匹配
       * @remark 仅支持字符串
       */
      Comparator["Like"] = "lk";
      /**
       * 字段存在
       */
      Comparator["Exist"] = "ex";
  })(exports.Comparator || (exports.Comparator = {}));
  (function (DataType) {
      /**
       * 全局数据
       */
      DataType[DataType["Global"] = 1] = "Global";
      /**
       * 事件数据
       */
      DataType[DataType["Event"] = 2] = "Event";
      /**
       * 网络数据
       */
      DataType[DataType["Network"] = 3] = "Network";
      /**
       * 页面上下文
       * @remark Native无此数据
       */
      DataType[DataType["Context"] = 4] = "Context";
  })(exports.DataType || (exports.DataType = {}));
  (function (DeviceType) {
      DeviceType["android"] = "android";
      DeviceType["iphone"] = "iphone";
      DeviceType["ipad"] = "ipad";
      DeviceType["windows"] = "windows";
      DeviceType["mac"] = "mac";
      DeviceType["others"] = "others";
  })(exports.DeviceType || (exports.DeviceType = {}));
  (function (ActType) {
      ActType["click"] = "clk";
      ActType["view"] = "view";
      ActType["vision"] = "vision";
      ActType["default"] = "";
  })(exports.ActType || (exports.ActType = {}));
  (function (NetworkStatusType) {
      NetworkStatusType["Wifi"] = "wifi";
      NetworkStatusType["N_5G"] = "5G";
      NetworkStatusType["N4G"] = "4G";
      NetworkStatusType["N3G"] = "3G";
      NetworkStatusType["N2G"] = "2G";
      NetworkStatusType["Others"] = "others";
  })(exports.NetworkStatusType || (exports.NetworkStatusType = {}));
  (function (KeyType) {
      /**
       * 字典键(dic) 按字典取值
       */
      KeyType["Dict"] = "d";
      /**
       * 数组键(arr) 按数组取值
       */
      KeyType["Arr"] = "a";
  })(exports.KeyType || (exports.KeyType = {}));
  (function (ValueType) {
      /**
       * 字符串
       */
      ValueType["String"] = "s";
      /**
       * 数字
       */
      ValueType["Number"] = "n";
  })(exports.ValueType || (exports.ValueType = {}));
  (function (BuiltInName) {
      BuiltInName["AppLaunch"] = "$AppLaunch";
      BuiltInName["AppShow"] = "$AppShow";
      BuiltInName["AppHide"] = "$AppHide";
      BuiltInName["PageShow"] = "$PageShow";
      BuiltInName["PageHide"] = "$PageHide";
      BuiltInName["PageShare"] = "$PageShare";
  })(exports.BuiltInName || (exports.BuiltInName = {}));

  // Import here Polyfills if needed. Recommended core-js (npm i -D core-js)
  var StatsCore = /** @class */ (function () {
      function StatsCore() {
      }
      return StatsCore;
  }());

  exports.default = StatsCore;

  Object.defineProperty(exports, '__esModule', { value: true });

})));
