// @zyb-data/stats-mp/dist/lib/mp/index.js
/// <reference path="../../typings/mp/index.d.ts" />
import { __assign } from "tslib";
import StatsMain from '../core';
import { StatsPlatformMp } from './platform';
import { D } from '../core/debug';
import { Env } from '../core/config';
import { version } from './version';
import { StatsAppFactoryWX } from './page';
var sdkVersion = version;
var logUrlDev = 'https://nlogtj.zuoyebang.cc/nlogtj/xiaochengxu_test';
var logUrlProd = 'https://nlogtj.zuoyebang.cc/nlogtj/xiaochengxu';
var ZybStatsWX = /** @class */ (function () {
    function ZybStatsWX() {
        this.cfg = {};
        this.sm = new StatsMain({ platform: new StatsPlatformMp() }, this.cfg);
        var appFactory = new StatsAppFactoryWX(this.sm);
        appFactory.init();
    }
    /**
     * 更新配置
     * @param config
     */
    ZybStatsWX.prototype.config = function (config) {
        // 小程序的上报地址控制收敛到小程序sdk入口这儿
        if (!config.logUrl) {
            switch (config.env) {
                case Env.Dev:
                    config.logUrl = logUrlDev;
                    break;
                case Env.Prod:
                    config.logUrl = logUrlProd;
                    break;
                default:
                    config.logUrl = logUrlProd;
                    break;
            }
        }
        this.cfg = __assign(__assign({}, this.cfg), config);
        this.sm.$config.set(config);
        this.sm.$dm.setGlobal({ plat: this.cfg.plat, zpID: this.cfg.zpID, sdkVersion: sdkVersion });
    };
    /**
     * 设置、更新全局变量
     * @param params
     */
    ZybStatsWX.prototype.global = function (params) {
        this.sm.$dm.setGlobal(params);
    };
    /**
     * 设置openid
     * @param openid
     */
    ZybStatsWX.prototype.setOpenid = function (openid, updateStore) {
        if (updateStore === void 0) { updateStore = true; }
        var _a;
        this.global({ i: openid });
        if (updateStore) {
            (_a = this.sm.$platform) === null || _a === void 0 ? void 0 : _a.setStore('openid', openid);
        }
    };
    /**
     * 设置unionid
     * @param unionid
     */
    ZybStatsWX.prototype.setUnionid = function (unionid, updateStore) {
        if (updateStore === void 0) { updateStore = true; }
        var _a;
        this.global({ mp_unionid: unionid });
        if (updateStore) {
            (_a = this.sm.$platform) === null || _a === void 0 ? void 0 : _a.setStore('unionid', unionid);
        }
    };
    ZybStatsWX.prototype.setAppid = function (appid, updateStore) {
        if (updateStore === void 0) { updateStore = true; }
        this.global({ mp_appid: appid });
    };
    /**
     * 设置登陆用户id（作业帮用户id，加密）
     * @param wxuid
     */
    ZybStatsWX.prototype.login = function (wxuid, updateStore) {
        if (wxuid === void 0) { wxuid = '-1'; }
        if (updateStore === void 0) { updateStore = true; }
        var _a;
        this.global({ paid: wxuid });
        if (updateStore) {
            (_a = this.sm.$platform) === null || _a === void 0 ? void 0 : _a.setStore('wxuid', wxuid);
        }
    };
    /**
     * 初始化
     */
    ZybStatsWX.prototype.init = function () {
        // 挂载到wx上
        wx.__zs = this;
        wx.zs = this;
        wx.zsd = D;
        D.log('全埋点SDK初始化');
        this.sm.init();
        var platform = this.sm.$platform;
        var openid = platform.getStore('openid');
        var unionid = platform.getStore('unionid');
        var wxuid = platform.getStore('wxuid');
        if (openid) {
            this.setOpenid(openid, false);
        }
        if (unionid) {
            this.setUnionid(unionid, false);
        }
        if (wxuid) {
            this.login(wxuid, false);
        }
    };
    /**
     * 主动埋点方法
     * @param name 点位名字
     * @param params 点位参数
     */
    ZybStatsWX.prototype.track = function (name, params) {
        return this.sm.track(name, params);
    };
    /**
     * 获取cuid
     */
    ZybStatsWX.prototype.getCuid = function () {
        var _a;
        return (_a = this.sm.$platform) === null || _a === void 0 ? void 0 : _a.getCUID();
    };
    return ZybStatsWX;
}());
export var zs = new ZybStatsWX();
