// sdk_cashier/types/argsConfig.ts
/*
 * @Author: Kong LiLi
 * @Date: 2019-06-19 16:30:43
 * @Last Modified by: Kong LiLi
 * @Last Modified time: 2019-07-15 17:06:23
 * model层的ts声明
 */
export interface PayArgs {
  orderId: string
  payChannel: number
  payRuleId: number
  transp?: string
  returnUrl?: string
  price?: number
  [prop: string]: any
}
export interface GetPayListArgs {
  mode: number
  appChannel: number
  source: string
  authServer: string
  sign: string
  orderId?: string
  subSource?: string
  // sdk补充参数
  payClient?: string
  [prop: string]: any
}

export interface GetSignListArgs {
  os: string
  vc?: string | number
  sdk?: string | number
  provider?: string
  source: string
  appChannel: string | number
  pid: string | number
  authServer: string | number
  sign: string | number
}

export interface PayAndSignArgs {
  payRuleId: string | number
  tplId: string | number
  notifyUrl?: string
  cnctNotifyUrl?: string
  username?: string
  uid: string | number
  source: string
  returnUrl: string
  firstWithholdTime?: string | number
  authServer: string
  orderId: string | number
  cuid: string | number
  transp: string | number
  appType: string | number
  appChannel: string | number
  payChannel: string | number
  wxAuthCode?: string | number
  os: string
  quitUrl: string
  contractCode: string
}

export interface JustSignArgs {
  tplId: string | number
  notifyUrl?: string
  cnctNotifyUrl?: string
  username?: string
  uid: string | number
  source: string | number
  returnUrl: string
  firstWithholdTime?: string | number
  appChannel: string | number
  userIp?: string
  contractCode: string
}
