// hybrid-sdk/src/index.ts
/*
 * @Author: liangpengfei
 * @Date: 2019/8/10 下午2:53:50
 * @Description: 程序入口
 */
import {logEntry, addLog} from './utils/tool'
import {jsBridgeInfo} from "hybrid-sdk-common"
// 显示log入口
// logEntry();
addLog('jsBridge UA开关:' + jsBridgeInfo.jsBridge_isNewJsBridge + '--|--系统版本:' + jsBridgeInfo.jsBridge_os_version + '--|--sdk版本:' + jsBridgeInfo.jsBridge_vc);

import * as business  from './business/index.js'
import * as core from  './core/index'

// import {getQueryString} from "./utils/tool"
// 预加载工具方法
export {transformCacheUrl} from "./utils/preload"

const newApi: boolean = +jsBridgeInfo.jsBridge_isNewJsBridge === 1 && jsBridgeInfo.jsBridge_vcnum >= 210;
const API: any = newApi ? core : business;
export {commonPay} from "./core/index"
export let {
    openWindow,
    // openOrderWindow,
    // openLiveWindow,
    goBack,
    getCommonParameters,
    share,
    login,
    checkIsLogin,
    // commonLog,
    // runtimeLog,
    // appConsoleLog,
    photoUpload,
    callPhoneKeyboard,
    download,
    copyToClipboard,
    playAudio,
    stopAudio,
    pauseAudioZyb,
    stopAudioZyb,
    showDialog,
    isWifi,
    request
}  = API;

export {isAppInstalled, hybrid, imgBrowser, openBrowser, windowConfig, removeNativeLoading} from "hybrid-sdk-common"
import * as commonSDK from "hybrid-sdk-common";
export let commonAPI = commonSDK;
export { hybrid as postMessageHybrid } from "./utils/hybrid"
export { hybrid as iframeHybrid } from "./utils/iframeHybrid"


export {
    checkIsInApp, checkSystem, getSystemInformation, getProductionName, getVersionCode, getVersionCodeName, isInOwnApp,
    // hybrid,
    // getCommonParameters, 
    scoreUpdate,
    commonLog, 
    liveNlog,
    requestDataFromNative,
    // request,
    timeSync,
    exitH5,
    startRecord, 
    endRecord, 
    playAudioZyb, 
    // pauseAudioZyb, 
    // stopAudioZyb, 
    // playAudio, 
    // stopAudio,
    exerciseFinish,
    liveCourseAnswerResult, 
    submitTestSuccess,
    // login, 
    // checkIsLogin,
    // download,
    setStorage, 
    getStorage,
    toast,
    enterLiveRoom, 
    enterPlayBackRoom,
    runtimeLog, 
    appConsoleLog,
    disableBack, 
    ableBack, 
    // goBack, 
    setBackMode,
    getLocation, 
    getUserinfo, 
    interceptSaveAddress,
    showCalendar,
    photoQuestion, 
    // photoUpload,
    chatWithRobot,
    commentReply, 
    teacherCommentDetail,
    reLoginDialog, 
    // showDialog,
    gotoPage, 
    navigate, 
    openWxApp,
    closeWebPage, 
    openWebPage, 
    disableClickWholeH5Page, 
    ableClickWholeH5Page,
    // commonPay, 
    rechargeCoin, 
    saveUserInformation, 
    redirectToPayPage, 
    showRealNameAlert,
    // callPhoneKeyboard,
    getSelectGrade,
    setInputPositionOnFocus,
    // copyToClipboard,
    setMemory,
    getMemory,
    commonJumpPage,
    gotoBindPhone,
    enterTest,
    // isWifi,
    pageLoadingComplete,
    liveOpenApp,
    openCustomerService, 
    openCachelist, 
    openSalelist, 
    openOrderWindow,
    // share,
    hideTitleBar, 
    showTitleBar, 
    editBarTitle,
    playVideo,
    disableWebScroll, 
    ableWebScroll,
    // openWindow, 
    openLiveWindow,
    listenNativeExit,
    listenUpdateData,
    listenVolumnMonitoring,
    listenGetWavetime,
    listenResultCallback,
    listenVisitKeyboard,
    listenLiveCourseAnswerResult,
    listenUpdateRefresh,
    listenPageEnter,
    listenPageLeave,
    listenPageDestory,
    on,
    off
} from './business'
