// hybrid-sdk/src/business/index.ts
// 系统相关配置 isInApp production、vc、vcname
export { checkIsInApp, checkSystem, getSystemInformation, getProductionName, getVersionCode, getVersionCodeName, isInOwnApp } from '../utils/system'

export { hybrid } from '../utils/iframeHybrid'
/* common api start  */
export { getCommonParameters, scoreUpdate } from './actions/commonActions/common'
export { commonLog, liveNlog } from './actions/commonActions/log'
export { requestDataFromNative } from './actions/commonActions/requestDataFromNative'
export { request } from './actions/commonActions/requst'
export { timeSync } from './actions/commonActions/timeSync'
export { exitH5 } from './actions/commonActions/exit'
export { startRecord, endRecord, playAudioZyb, pauseAudioZyb, stopAudioZyb, playAudio, stopAudio } from './actions/commonActions/audio'
export { exerciseFinish } from './actions/commonActions/dot'

export { liveCourseAnswerResult, submitTestSuccess } from './actions/commonActions/score'

export { login, checkIsLogin } from './actions/commonActions/login'

export { download } from './actions/commonActions/download'

export { setStorage, getStorage } from './actions/commonActions/storage'

export { toast } from './actions/commonActions/toast'

export { enterLiveRoom, enterPlayBackRoom } from './actions/commonActions/liveAndPalyBackRoom'

/** app api  start */
export { runtimeLog, appConsoleLog } from './actions/appActions/appLog'

export { disableBack, ableBack, goBack, setBackMode } from './actions/appActions/back'

export { getLocation, getUserinfo, interceptSaveAddress } from './actions/appActions/basicInformation'

export { showCalendar } from './actions/appActions/calendar'

export { photoQuestion, photoUpload } from './actions/appActions/camera'

export { chatWithRobot } from './actions/appActions/chat'
export { commentReply, teacherCommentDetail } from './actions/appActions/comment'

export { reLoginDialog, showDialog } from './actions/appActions/dialog'

export { gotoPage, navigate, openWxApp } from './actions/appActions/go'

export { closeWebPage, openWebPage, disableClickWholeH5Page, ableClickWholeH5Page } from './actions/appActions/h5plugin'

export { commonPay, rechargeCoin, saveUserInformation, redirectToPayPage, showRealNameAlert } from './actions/appActions/live'

export {
  callPhoneKeyboard,
  getSelectGrade,
  setInputPositionOnFocus,
  copyToClipboard,
  setMemory,
  getMemory,
  commonJumpPage,
  gotoBindPhone,
  enterTest,
  isWifi,
  pageLoadingComplete,
  liveOpenApp
} from './actions/appActions/other'

export { openCustomerService, openCachelist, openSalelist, openOrderWindow } from './actions/appActions/service'

export { share } from './actions/appActions/share'

export { hideTitleBar, showTitleBar, editBarTitle } from './actions/appActions/titleBar'

export { playVideo } from './actions/appActions/vedio'

export { disableWebScroll, ableWebScroll } from './actions/appActions/webScroll'

export { openWindow, openLiveWindow } from './actions/appActions/window'

export {
  listenNativeExit,
  listenUpdateData,
  listenVolumnMonitoring,
  listenGetWavetime,
  listenResultCallback,
  listenVisitKeyboard,
  listenLiveCourseAnswerResult,
  listenUpdateRefresh,
  listenPageEnter,
  listenPageLeave,
  listenPageDestory,
  on,
  off
} from './subscribe/index'
