// hybrid-sdk-common/src/core/openWindow.ts
import { hybrid } from "../utils/hybrid";
import { getMergedCallback, getPromiseFunc} from "../utils/tool";

interface openWindowParams {
    pageUrl: string;
    title?: string;
    swapBack?: number;
    hideStatusBar?: number;
    hideNavBar?: number;
    screenDirection?: number;
    navBarBgColor?: string;
    staBarStyle?: number;
    showShareBtn?: number;
    allLight?: number;
    businessData?: any;
    [propName: string]: any;
}
const win: any = window;
const winClassName:any = (win.getOpenWindowClassName && win.getOpenWindowClassName()) || null;
const defaultParams: openWindowParams = {
    pageUrl: "",                 // 页面地址
    title: "",                   // 页面标题
    swapBack: 0,                 // 滑动手势返回 0禁用 1允许
    hideStatusBar: 0,            // 隐藏状态栏 0显示 1不显示
    hideNavBar: 0,               // 隐藏导航栏（安卓:titleBar） 0显示 1不显示
    screenDirection: 0,          // 屏幕方向 0竖屏 1横屏 2跟随设备方向
    navBarBgColor: "",           // 导航条背景色
    staBarStyle: 0,              // 状态栏风格 0白色 1黑色
    showShareBtn: 0,             // 是否显示右上角分享按钮<样式固定>
    allLight: 0,                 // 屏幕是否常亮 1 常亮 0跟随系统
    // businessData: winClassName// 扩展参数 
};

const containerType = ["",
    'WebViewController', // 作业帮 ios
    'com.baidu.homework.activity.web.ZybWebActivity', // 作业帮 android
    'ZYBWKWebViewController', // 一课 ios
    'com.baidu.homework.activity.live.web.LiveCacheHybridActivity' // 一课 android
];

/**
 * 在一个新的nativepage中新开一个页面
 * @param params 新window的属性
 * @param params.businessData.className_android activity的类名，例如：com.baidu.homework.activity.web.ZybWebActivity
 * @param params.businessData.className_ios 容器的类名，例如：xx.xxx.xxxx
 * @param callback 
 */
export function core_openWindow(params: any){
    if(params.businessData) {
        let classType = params.businessData.classType;
        if(classType.ios) {
            params.businessData.className_ios = containerType[classType.ios];
        }
        if(classType.android) {
            params.businessData.className_android = containerType[classType.android];
        }
        delete params.businessData.classType;
        params.businessData = {
            ...winClassName,
            ...params.businessData,
        }
    }else {
        params.businessData = winClassName;
    }
    let _params: openWindowParams = {
        ...defaultParams,
        ...params
    };
    
    return getPromiseFunc("core_openWindow", _params);
}

interface windowConfigParams {
    pageUrl?: string;             //页面地址
    title?: string;               //页面标题
    swapBack?: number;            //滑动手势返回 0禁用 1允许
    hideStatusBar?: number;       //隐藏状态栏 0显示 1不显示
    hideNavBar?: number;          //隐藏导航栏 0显示 1不显示
    screenDirection?: number;     //屏幕方向 0竖屏 1横屏 2跟随设备方向
    navBarBgColor?: string;       //导航条背景色 //android不支持
    staBarStyle?: number;         //状态栏风格 0白色 1黑色 //android不支持
    allLight?: number;            //屏幕是否常亮 0: 跟随系统设置 1: 常亮
    showShareBtn?:  number;       //是否显示右上角分享按钮 0 不显示 1 显示
    shareBtnBgImg?: string;       //分享按钮背景图
    shareData?: object;           //参数同 core_share 
    shareCallBack?: Function;     //分享回调；
    backShowDialog?: number;      //点击返回按钮是否展示弹窗 0 show 1 不show
    dialogData?: object;          //参数格式和core_showDialog一致
    dialogCallback?: Function;    //dialog回调，回调数据会指出用户点击了哪个按钮，默认左按钮留下，右按钮离开
    businessData?: any;
    [propName: string]: any;
}

const defaultWindowConfigParams: windowConfigParams = {
    pageUrl: "",                 // 页面地址
    title: "",                   // 页面标题
    swapBack: 0,                 // 滑动手势返回 0禁用 1允许
    hideStatusBar: 0,            // 隐藏状态栏 0显示 1不显示
    hideNavBar: 0,               // 隐藏导航栏（安卓:titleBar） 0显示 1不显示
    screenDirection: 0,          // 屏幕方向 0竖屏 1横屏 2跟随设备方向
    navBarBgColor: "",           // 导航条背景色
    staBarStyle: 0,              // 状态栏风格 0白色 1黑色
    allLight: 0,                 // 屏幕是否常亮 1 常亮 0跟随系统
    showShareBtn: 0,             // 是否显示右上角分享按钮<样式固定>
    shareBtnBgImg: '',           //分享按钮背景图
    shareData: {},               //参数同 core_share 
    shareCallBack: function(){}, //分享回调；
    backShowDialog: 0,           //点击返回按钮是否展示弹窗 0 show 1 不show
    dialogData: {},              //参数格式和core_showDialog一致
    dialogCallback: function(){
        getPromiseFunc("core_exit", {backWindow: 1});
    },//dialog回调，回调数据会指出用户点击了哪个按钮，默认左按钮留下，右按钮离开
    // businessData: null        //  预留参数（上传业务层）
};


export function windowConfig(params: any){
    let _params = {
        // ...defaultWindowConfigParams,
        ...params
    };

    return new Promise(function(resolve: Function, reject: Function){
        let success = function(res: any){
            resolve(res);
        };
    
        let fail = function(res: any){
            reject(res);
        };

        let _callback = function(res: any){
            let defaultCallback = getMergedCallback(success, fail);
            let shareCallBack = _params.shareCallBack || defaultWindowConfigParams.shareCallBack;
            let dialogCallback = _params.dialogCallback || defaultWindowConfigParams.dialogCallback;
    
            if(res.data && res.data.result+1 > 0) {
                shareCallBack(res.data);
            }else if(res.data && res.data.res){
                dialogCallback(res.data);
            }else{
                defaultCallback(res.data);
            }
        };
        hybrid('core_windowConfig', _params, _callback);
    });

    
}


/**
 * 在端外浏览器打开url
 * @param url 要打开的地址 
 */
export function openBrowser(params: any){
    hybrid('core_openBrowser', params);
}

/**
 * 页面的退出方式
 * @export
 * @param {number} backWindow 退出方式 0 fe页面一级级返回;   1  native 结束
 */
export function core_goBack(params: {backWindow: 0 | 1;}) {
    let _params: any = {
        backWindow: 0,
        ...params
    };

    return getPromiseFunc("core_exit", _params);

}
