// hybrid-sdk-common/src/core/openUrl.ts
import { getPromiseFunc } from "../utils/tool";

interface openUrlParams {
    module: string,
    page: string,
    params: string,
    title?: string;
    swapBack?: number;
    hideStatusBar?: number;
    hideNavBar?: number;
    screenDirection?: number;
    navBarBgColor?: string;
    staBarStyle?: number;
    showShareBtn?: number;
    allLight?: number;
    businessData?: any;
    [propName: string]: any;
}

const containerType = ["",
    'WebViewController', // 作业帮 ios
    'com.baidu.homework.activity.web.ZybWebActivity', // 作业帮 android
    'ZYBWKWebViewController', // 一课 ios
    'com.baidu.homework.activity.live.web.LiveCacheHybridActivity' // 一课 android
];

const win: any = window;
const winClassName:any = (win.getOpenWindowClassName && win.getOpenWindowClassName()) || null;

/**
 * 
 * @param params 打开native 或 web页面
 */
export function core_openUrl(params: any){  
    if(params.businessData) {
        let classType = params.businessData.classType;
        if(classType.ios) {
            params.businessData.className_ios = containerType[classType.ios];
        }
        if(classType.android) {
            params.businessData.className_android = containerType[classType.android];
        }
        delete params.businessData.classType;
        params.businessData = {
            ...winClassName,
            ...params.businessData,
        }
    }else {
        params.businessData = winClassName;
    }
    let _params: openUrlParams = {
        ...params
    };  
    return getPromiseFunc("core_openUrl", _params);
}
