// hybrid-sdk/src/core/payToy.ts
import {resolveCallbackFactory, rejectCallbackFactory, getIframePromiseFunc} from "../utils/tool";
import {commonPay as core_commonPay, hybrid, jsBridgeInfo} from  'hybrid-sdk-common'

interface commonPayParams {
    channel: number ;  // 渠道分类 http://wiki.afpai.com/pages/viewpage.action?pageId=19902070
    info: object; // 协议内容，调用第三方支付需要的信息，比如，商户号，通知地址，APPID
    source?: number; // 支付来源，比如，订单列表，需要和端协商来定义
}
interface commonPayResponse {
    status: number;
    errstr: string;
}

/**
 * 监测应用是否安装
 * @param params 
 */
const newApi: boolean = +jsBridgeInfo.jsBridge_isNewJsBridge === 1 && jsBridgeInfo.jsBridge_vcnum >= 210;
export function commonPay(params: commonPayParams, callback): Promise<any> {
    let success= resolveCallbackFactory(callback);
    let fail= rejectCallbackFactory(callback);
    if(newApi){
        return core_commonPay(params).then(success, fail);
    }else {
        // hybrid('backup_FIRE', params, callback);
        return getIframePromiseFunc('backup_FIRE', params).then(success, fail); 
    }
}
