// @zyb-data/stats-mp/dist/lib/core/debug.js
import { __spreadArrays } from "tslib";
var StatsDebugger = /** @class */ (function () {
    function StatsDebugger() {
        this.logsCache = [];
        this.isDebug = true;
        this.isInfo = true;
    }
    StatsDebugger.prototype.switch = function (isDebug) {
        this.isDebug = isDebug;
    };
    StatsDebugger.prototype.group = function (label) {
        if (this.isDebug) {
            console.group('%c _ZS_ ' + label, 'background: #666; color: #badaff');
        }
    };
    StatsDebugger.prototype.groupEnd = function () {
        if (this.isDebug) {
            console.groupEnd();
        }
    };
    StatsDebugger.prototype.log = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        if (this.isDebug) {
            console.log.apply(console, __spreadArrays(['%c _ZS_LOG_ ', 'background: #666; color: #badaff'], args));
        }
    };
    StatsDebugger.prototype.info = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        if (this.isInfo) {
            console.log.apply(console, __spreadArrays(['%c _ZS_INF_ ', 'background: #666; color: #bada55'], args));
        }
    };
    StatsDebugger.prototype.sys = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        console.log.apply(console, __spreadArrays(['%c _ZS_SYS_ ', 'background: #666; color: #ffffff'], args));
    };
    StatsDebugger.prototype.error = function () {
        var args = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            args[_i] = arguments[_i];
        }
        console.log.apply(console, __spreadArrays(['%c _ZS_ERR_ ', 'background: #666; color: #ff5555'], args));
    };
    StatsDebugger.prototype.pushRecord = function (log) {
        this.logsCache.push(log);
    };
    StatsDebugger.prototype.clearRecords = function (log) {
        this.logsCache = [];
    };
    StatsDebugger.prototype.groupRecords = function () {
        console.log('%c _ZS_RECORDS_GROUP_ ', 'background: #666; color: #bada55');
        console.group('已打点位');
        this.logsCache.forEach(function (v, i) {
            console.log(i, v.name, v.lnc, v.pageUUID);
        });
        console.groupEnd();
    };
    StatsDebugger.prototype.tableRecords = function () {
        console.log('%c _ZS_RECORDS_TABLE_ ', 'background: #666; color: #bada55');
        console.table(this.logsCache, ['name', 'lnc', 'pageUUID']);
    };
    return StatsDebugger;
}());
export var D = new StatsDebugger();
