// hybrid-sdk-common/src/index.ts
export {
    core_goBack,
    goBack,
    openBrowser,
    windowConfig,
    core_openWindow,
    openWindow,
    core_share,
    share,
    checkLogin,
    core_login,
    login,
    nativeLog,
    fetchImg,
    dial,
    downloader,
    copyText,
    imgBrowser,
    core_audio,
    audio,
    core_showDialog,
    showDialog,
    netType,
    core_request,
    request,
    core_getCommonParameters,
    getCommonParameters,
    isAppInstalled,
    removeNativeLoading,
    pageTo,
    hybrid,
    commonPay,
    setStorage, 
    getStorage
} from "./core/index"

export {
    NativeCallBackData, // interface
    ActionType,         // enum
    onPageEnter,
    onPageLeave,
    onPageDestroy,
    onPagePullFresh,
    onNativeCallback  // 自定义全局回调
} from "./hook/index"

export { transformCacheUrl } from "./utils/preload"
export { switchConfig as jsBridgeInfo, canUseCommon } from "./utils/system"
export { hybrid as postMessageHybrid } from "./utils/hybrid"
export { hybrid as iframeHybrid } from "./utils/iframeHybrid"
