// hybrid-sdk/src/business/actions/appActions/dialog.ts
import { hybrid } from '../../../utils/iframeHybrid'
import Type from '../../../utils/type'

/**
 * http://qatest7.suanshubang.com/static/fe-docs/_book/hybrid-sdk/app_dialog.html
 */

/**
 * @description 异地登录了,你的账号已在别处登录,点击重新登录后可继续使用.如果登录失败，是跳转到登录页面
 * @param {Object}  options
 */
export function reLoginDialog (data: any, callback: Function) {
  var { needCallBack } = data
  if (needCallBack !== 0 && needCallBack !== 1) {
    console.error('errstr is undefined')
    return
  }

  if (needCallBack === 0 && !Type.isFunction(callback)) {
    callback = function () { }
  }

  if (!Type.isFunction(callback)) {
    console.error('errstr is undefined')
    return
  }
  hybrid('reLoginDialog', data, callback)
}

/**
 * @description 展示提示框
 * @param {Object} options
 */
export function showDialog (data: any, callback: Function) {
  var {
    description,
    neutralText,
    negativeText,
    positiveText,
    cancelOutside
  } = data
  var params = null
  if (neutralText) {
    params = {
      description: description,
      neutralText: neutralText,
      cancelOutside: cancelOutside
    }
  } else {
    params = {
      description: description,
      negativeText: negativeText,
      positiveText: positiveText,
      cancelOutside: cancelOutside
    }
  }
  hybrid('showDialog', params, callback)
}
