// hybrid-sdk/src/business/actions/appActions/window.ts
import { hybrid } from '../../../utils/iframeHybrid'

// http://qatest7.suanshubang.com/static/fe-docs/_book/hybrid-sdk/app_window.html

/**
 * @description push新的webviewController
 * @param {Object} options
 */
export function openWindow (data: any, callback: Function) {
  var { url, dialogTitle, dialogSubTitle, positiveText, negativeText, dialogCloseBtn, hideNav } = data
  if (!url) {
    console.error('url is required')
    return false
  }

  var parameters = {
    url: url,
    dialogTitle: dialogTitle,
    dialogSubTitle: dialogSubTitle,
    dialogLeftText: positiveText,
    dialogRightText: negativeText,
    dialogCloseBtn: dialogCloseBtn,
    hideNav: hideNav
  }

  hybrid('openWindow', parameters, callback)
}

/**
 * @description 直播开一个h5的web页面
 * @param {*} data
 * @param {*} callback
 */
export function openLiveWindow (data: any, callback: Function) {
  if (!data || !data.url) {
    console.error('url is required')
    return false
  }

  hybrid('openLiveWindow', data, callback)
}
