// @yaya/director-uni/lib/yaya-scene-quiz-in-class/questions/js/audioSdkInstance.js
import { EventEmitter } from "../../../yaya-common/event-emitter";
import voiceSdk from "../../../yaya-common/audioSdk";
/**
 * 单句跟读/古诗背诵/开放口述题型 voiceSDK实例
 */
class vSdk extends EventEmitter {
  constructor() {
    super();
    this.vSdk = null;
  }

  start(params, isReStart) {
    console.log("_events_events_events1111111", this._events);
    // 已有实例时，销毁EventEmitter实例
    // if (isReStart) {
    //   this.off("result", this._events["result"][0]);
    // }
    console.log("_events_events_events2222222", this._events);
    // 初始化新实例并启动
    this.vSdk = new voiceSdk(params);
    this.vSdk.start();
    this.initAction();
  }
  resume(params) {
    // 初始化新实例并启动
    this.vSdk.resume(params);
  }

  // 销毁voiceSdk实例
  destroySdk() {
    console.log("销毁voiceSdk实例");
    this.vSdk && this.vSdk.destory();
  }

  stop(cb) {
    this.vSdk && this.vSdk.stop(cb);
  }

  pause() {
    this.vSdk && this.vSdk.pause();
  }

  initAction() {
    this.vSdk.onError(() => {
      this.emit("error");
    });
  }

  // voiceSDK实时结果事件
  onResult(cb) {
    // this.on("result", cb);
    this.vSdk.onResult((res) => {
      console.log("audioSdkInstance  onresult", res);
      cb && cb(res);
      // this.emit("result", res);
    });
  }

  // voiceSDK错误事件
  onError(handler) {
    this.on("error", handler);
  }
}

export default new vSdk();
