// hybrid-sdk/src/utils/system.ts
/**
 * Copyright (c) 2014-2018 Zuoyebang, All rights reserved.
 * @fileoverview 作业帮 端的类型，系统，vc版本号，vcname版本号信息.
 * @author liujianwei | liujianwei@zuoyebang.com
 * @version 1.0 | 2018-10-03 | liujianwei    // 初始版本。
 *
 * @return                               // 对像
 *   @property isInApp {Boolean}         // 属性：app则为true，window端为false
 *   @property system {String}           // 属性：端的系统类型，iPad,iPhone,Android
 *   @property production {String}       // 属性：一课：ariclass，作业帮：homework
 *   @property vc{String}                // 属性：产品的vc版本号
 *   @property vcname{String}            // 属性：产品的vcname版本号
 *
 * @description    // 系统信息。
 *   1) 用于了解系统的基本信息
 *
 * @example    // 典型的调用示例。
     import system from '@/utils/system.js';
 */

// 检测平台是否为app
const userAgent: any = navigator.userAgent.toLowerCase();
// 浏览器类型
const IPHONE = 'iphone';
const IPAD = 'ipad';
const ANDROID = 'android';
const WINDOWS = 'windows';
const browser = {
  android: userAgent.indexOf(ANDROID) > -1 || userAgent.indexOf('adr') > -1, // android终端
  iPhone: userAgent.indexOf(IPHONE) > -1, // 是否为iPhone或者QQHD浏览器
  iPad: userAgent.indexOf(IPAD) > -1 // 是否iPad
}

// 检测系统类型
export const _isInApp = (function () {
  return browser.android || browser.iPhone || browser.iPad
})()

// 系统类型
export const system = (function () {
  if (browser.iPhone) {
    return IPHONE
  } else if (browser.iPad) {
    return IPAD
  } else if (browser.android) {
    return ANDROID
  } else {
    return WINDOWS
  }
})()
// 产品列表
const AIRCLASS = _isInApp ? 'airclass' : 'a_irclass' // 一课
const HOMEWORK = _isInApp ? 'homework' : 'h_omework' // 作业帮
export const UNKNOW = 'unknow' // 未知

// 产品类型
export const production = (function () {
  if (userAgent.indexOf(AIRCLASS) > -1) {
    return AIRCLASS
  } else if (userAgent.indexOf(HOMEWORK) > -1) {
    return HOMEWORK
  } else {
    return UNKNOW
  }
})()

// trim
export function myTrim (x: any) {
  return x.replace(/^\s+|\s+$/gm, '')
}

// vc版本号
export const vc = (function () {
  var vcType = `${production}_vc/` // ariclass_vc/
  var regex = `${vcType}[\\d.]+\\s\*`
  var regVc = new RegExp(regex, 'ig') /// airclass_vc\/[\d.]+\s/ig
  var vcArray = userAgent.match(regVc)
  if (Array.isArray(vcArray)) {
    return myTrim(vcArray[0].replace(vcType, ''))
  }
  return ''
})()

// vcname版本号
export const vcname = (function () {
  var vcNameType = `${production}_vcname/` // ariclass_vc/
  var regVcname = new RegExp(`${vcNameType}[\\d.]+\\s\*`, 'ig') /// airclass_vcname\/[\d.]+\s/ig
  var vcnameArray = userAgent.match(regVcname)
  if (Array.isArray(vcnameArray)) {
    return myTrim(vcnameArray[0].replace(vcNameType, ''))
  }
  return ''
})()

// 输出系统类型和系统名称

export const isInApp = _isInApp
export function checkIsInApp () {
  return _isInApp
}
/**
 * 注释
 */
export function getSystemInformation () {
  return system
}

export function getProductionName () {
  return production.replace('_', '')
}

export function getVersionCode () {
  return vc
}

export function getVersionCodeName () {
  return vcname
}
export function isInOwnApp () {
  // if (_isInApp) {
  //   return production.replace('_', '') !== UNKNOW
  // } else {
  //   return false
  // }
  return production.replace('_', '') !== UNKNOW
}

export var checkSystem = {
  isAndroid: function () {
    return browser.android
  },
  isIPhone: function () {
    return browser.iPhone
  },
  isIPad: function () {
    return browser.iPad
  }
}

export const switchConfig: any = (function() {
  let uaObj: any = {};
  let uaArr = window.navigator.userAgent.split(' ');
  uaArr = uaArr.filter((item)=>{
    return item.indexOf('jsBridge_isNewJsBridge/') >= 0 || item.indexOf('jsBridge_vc/') >= 0 || item.indexOf('jsBridge_os_version/') >= 0;
  });

  uaArr.forEach((item)=>{
    const arr = item.split('/');
    uaObj[arr[0]]=arr[1];
  });
  if(uaObj.jsBridge_vc){
    let vc = uaObj.jsBridge_vc;
    vc = +(vc.split('.').join(''));
    uaObj.jsBridge_vcnum = vc;
  }
  return uaObj;
})();
