// @zyb-data/stats-core/dist/lib/model/common.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * 事件类型枚举
 */
var EventDataType;
(function (EventDataType) {
    /**
     * 进入页面
     */
    EventDataType["pageIn"] = "pageIn";
    /**
     * 离开页面
     */
    EventDataType["pageOut"] = "pageOut";
    /**
     * 控件显示
     */
    EventDataType["viewShow"] = "viewShow";
    /**
     * 控件隐藏
     */
    EventDataType["viewHide"] = "viewHide";
    /**
     * 点击
     */
    EventDataType["click"] = "click";
    /**
     * 请求开始
     */
    EventDataType["reqStart"] = "reqStart";
    /**
     * 请求成功
     */
    EventDataType["reqSucc"] = "reqSucc";
    /**
     * 请求失败
     */
    EventDataType["reqFail"] = "reqFail";
    /**
     * 分享
     */
    EventDataType["share"] = "share";
})(EventDataType = exports.EventDataType || (exports.EventDataType = {}));
/**
 * 比较符枚举
 * @remark 未作特殊说明的操作符，对任意类型都支持
 */
var Comparator;
(function (Comparator) {
    /**
     * 等于
     */
    Comparator["Equal"] = "eq";
    /**
     * 不等于
     */
    Comparator["NEQ"] = "neq";
    /**
     * 大于
     * @remark 仅支持数字
     */
    Comparator["GT"] = "gt";
    /**
     * 小于
     * @remark 仅支持数字
     */
    Comparator["LT"] = "lt";
    /**
     * 字符串匹配
     * @remark 仅支持字符串
     */
    Comparator["Like"] = "lk";
    /**
     * 字段存在
     */
    Comparator["Exist"] = "ex";
})(Comparator = exports.Comparator || (exports.Comparator = {}));
/**
 * 资源类型枚举
 */
var DataType;
(function (DataType) {
    /**
     * 全局数据
     */
    DataType[DataType["Global"] = 1] = "Global";
    /**
     * 事件数据
     */
    DataType[DataType["Event"] = 2] = "Event";
    /**
     * 网络数据
     */
    DataType[DataType["Network"] = 3] = "Network";
    /**
     * 页面上下文
     * @remark Native无此数据
     */
    DataType[DataType["Context"] = 4] = "Context";
})(DataType = exports.DataType || (exports.DataType = {}));
/**
 * 设备类型枚举
 */
var DeviceType;
(function (DeviceType) {
    DeviceType["android"] = "android";
    DeviceType["iphone"] = "iphone";
    DeviceType["ipad"] = "ipad";
    DeviceType["windows"] = "windows";
    DeviceType["mac"] = "mac";
    DeviceType["others"] = "others";
})(DeviceType = exports.DeviceType || (exports.DeviceType = {}));
var ActType;
(function (ActType) {
    ActType["click"] = "clk";
    ActType["view"] = "view";
    ActType["vision"] = "vision";
    ActType["default"] = "";
})(ActType = exports.ActType || (exports.ActType = {}));
/**
 * 网络状态枚举
 */
var NetworkStatusType;
(function (NetworkStatusType) {
    NetworkStatusType["Wifi"] = "wifi";
    NetworkStatusType["N_5G"] = "5G";
    NetworkStatusType["N4G"] = "4G";
    NetworkStatusType["N3G"] = "3G";
    NetworkStatusType["N2G"] = "2G";
    NetworkStatusType["Others"] = "others";
})(NetworkStatusType = exports.NetworkStatusType || (exports.NetworkStatusType = {}));
/**
 * 键类型枚举
 */
var KeyType;
(function (KeyType) {
    /**
     * 字典键(dic) 按字典取值
     */
    KeyType["Dict"] = "d";
    /**
     * 数组键(arr) 按数组取值
     */
    KeyType["Arr"] = "a";
})(KeyType = exports.KeyType || (exports.KeyType = {}));
/**
 * 值类型枚举
 */
var ValueType;
(function (ValueType) {
    /**
     * 字符串
     */
    ValueType["String"] = "s";
    /**
     * 数字
     */
    ValueType["Number"] = "n";
})(ValueType = exports.ValueType || (exports.ValueType = {}));
/**
 * 内置事件名枚举
 */
var BuiltInName;
(function (BuiltInName) {
    BuiltInName["AppLaunch"] = "$AppLaunch";
    BuiltInName["AppShow"] = "$AppShow";
    BuiltInName["AppHide"] = "$AppHide";
    BuiltInName["PageShow"] = "$PageShow";
    BuiltInName["PageHide"] = "$PageHide";
    BuiltInName["PageShare"] = "$PageShare";
})(BuiltInName = exports.BuiltInName || (exports.BuiltInName = {}));
